Pull DiagramOutputPort into its own file and untangle from Diagram.

sherm1 opened the pull request:
This is a small cleanup PR to extract DiagramOutputPort from its hiding place, untangle it from Diagram, and generally make it parallel to LeafOutputPort.

The one remaining hidden class in diagram.h is gone in the caching branch, after which diagram.h will contain just Diagram.

Reviewers: DiagramOutputPort code was moved intact to diagram_output_port.h, then mildly modified as @jwnimmer-tri did to LeafOutputPort in #8786 so that it does not depend on Diagram.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8857)
<!-- Reviewable:end -->

sherm1 commented:
+@jwnimmer-tri for feature review please.
+@seancurtis-tri for platform review per rotation please.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8857#-:-LDA4gT44YKDQH1J6ndA:b9qvilp)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[systems/framework/BUILD.bazel, line 490 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDC-L-C83EmOeLdma40:-LDC-L-D1Pa5s3NvOpoS:b-u7z9w7) ([raw file](https://github.com/robotlocomotion/drake/blob/a4e8513ecfaaf2013fe2ee567b2c6b77f431c8c5/systems/framework/BUILD.bazel#L490)):*
> ```Python
>         ":output_port",
>         ":value",
>         ":vector",
> ```

BTW unused?

---

*[systems/framework/BUILD.bazel, line 493 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDC-DErBEbdP6_m8cv1:-LDC-DEs-mjPnzFOguOb:b-u7z9w7) ([raw file](https://github.com/robotlocomotion/drake/blob/a4e8513ecfaaf2013fe2ee567b2c6b77f431c8c5/systems/framework/BUILD.bazel#L493)):*
> ```Python
>         "//common:default_scalars",
>         "//common:essential",
>         "//common:nice_type_name",
> ```

BTW unused?

---

*[systems/framework/diagram_output_port.h, line 30 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDC-zsM9PckYZWuIwm7:-LDC-zsM9PckYZWuIwm8:b2cblzf) ([raw file](https://github.com/robotlocomotion/drake/blob/a4e8513ecfaaf2013fe2ee567b2c6b77f431c8c5/systems/framework/diagram_output_port.h#L30)):*
> ```objc
>   /// indicated port. That port's owning system must be a subsystem of the
>   /// diagram.
>   DiagramOutputPort(const System<T>& diagram,
> ```

Either this argument should be typed as `const Diagram<T>&`, or else the forward declaration on line 16 is unused.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8857#-:-LDC0Oas3KCEtuQccqV-:bnfp4nl)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with some cosmetic updates.

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[systems/framework/diagram_output_port.h, line 29 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDCIk_H4ECAmZUoUf70:-LDCIk_H4ECAmZUoUf71:b-w7kbqv) ([raw file](https://github.com/robotlocomotion/drake/blob/a4e8513ecfaaf2013fe2ee567b2c6b77f431c8c5/systems/framework/diagram_output_port.h#L29)):*
> ```objc
>   /// Construct a %DiagramOutputPort for the given `diagram` that exports the
>   /// indicated port. That port's owning system must be a subsystem of the
>   /// diagram.
> ```

BTW The constructor signature changed, but not the documentation. Particularly with respect to the diagram and system_base. This has come out of the internal namespace so it deserves the full treatment.

---

*[systems/framework/diagram_output_port.h, line 56 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDCKTaw33Df_t2-wQit:-LDCKTaw33Df_t2-wQiu:bnuwstx) ([raw file](https://github.com/robotlocomotion/drake/blob/a4e8513ecfaaf2013fe2ee567b2c6b77f431c8c5/systems/framework/diagram_output_port.h#L56)):*
> ```objc
>   }
> 
>   // Pass in just the source System's Context, not the whole Diagram context
> ```

BTW this comment feels misleading. The word "pass" outside the function suggests a description of the input `context` parameter. However, the call to `get_subtext` seems to contradict it.

---

*[systems/framework/diagram_output_port.h, line 59 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDCLjbg7VAdrJcukacZ:-LDCLjbg7VAdrJcukac_:b-ibo9na) ([raw file](https://github.com/robotlocomotion/drake/blob/a4e8513ecfaaf2013fe2ee567b2c6b77f431c8c5/systems/framework/diagram_output_port.h#L59)):*
> ```objc
>   // we're given.
>   void DoCalc(
>       const Context<T>& context, AbstractValue* value) const final {
> ```

FYI I suspect this layout wouldn't survive a clang format (and isn't necessary).

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8857#-:-LDCP7FU0l-7IVKtePNR:b-1mfvyk)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Cleaned up and documented. All comments addressed, PTAL.

---

Review status: 4 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[systems/framework/BUILD.bazel, line 490 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDC-L-C83EmOeLdma40:-LDCdd2RD03d3OiUjQ-6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a4e8513ecfaaf2013fe2ee567b2c6b77f431c8c5/systems/framework/BUILD.bazel#L490)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW unused?
</blockquote></details>

Done.

---

*[systems/framework/BUILD.bazel, line 493 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDC-DErBEbdP6_m8cv1:-LDCddd42aNyUzQhkHVX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a4e8513ecfaaf2013fe2ee567b2c6b77f431c8c5/systems/framework/BUILD.bazel#L493)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW unused?
</blockquote></details>

Done.

---

*[systems/framework/diagram_output_port.h, line 29 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDCIk_H4ECAmZUoUf70:-LDCs-DB3gUEBZqEP-Sm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a4e8513ecfaaf2013fe2ee567b2c6b77f431c8c5/systems/framework/diagram_output_port.h#L29)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW The constructor signature changed, but not the documentation. Particularly with respect to the diagram and system_base. This has come out of the internal namespace so it deserves the full treatment.
</blockquote></details>

Done.

---

*[systems/framework/diagram_output_port.h, line 30 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDC-zsM9PckYZWuIwm7:-LDCdhU9BmbHZG31C9Ja:b-n6zu4t) ([raw file](https://github.com/robotlocomotion/drake/blob/a4e8513ecfaaf2013fe2ee567b2c6b77f431c8c5/systems/framework/diagram_output_port.h#L30)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Either this argument should be typed as `const Diagram<T>&`, or else the forward declaration on line 16 is unused.
</blockquote></details>

Done. Deleted the forward decl. (Has to be System because now this class can't downcast from Diagram to System.)

---

*[systems/framework/diagram_output_port.h, line 56 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDCKTaw33Df_t2-wQit:-LDCg2Yt0zkuIRnK-SVZ:bo9dn4j) ([raw file](https://github.com/robotlocomotion/drake/blob/a4e8513ecfaaf2013fe2ee567b2c6b77f431c8c5/systems/framework/diagram_output_port.h#L56)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW this comment feels misleading. The word "pass" outside the function suggests a description of the input `context` parameter. However, the call to `get_subtext` seems to contradict it.
</blockquote></details>

Done. Improved (hopefully) formatting and documentation for all the private methods, ptal. 

---

*[systems/framework/diagram_output_port.h, line 59 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDCLjbg7VAdrJcukacZ:-LDCe8hF0miNXw8PxLC9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a4e8513ecfaaf2013fe2ee567b2c6b77f431c8c5/systems/framework/diagram_output_port.h#L59)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI I suspect this layout wouldn't survive a clang format (and isn't necessary).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8857#-:-LDCs6hm3KYgO_utMmNN:b-ugv7ko)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8857)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Almost...

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[systems/framework/diagram_output_port.h, line 36 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDCsuTV4fGvTETCEAiM:-LDCsuTV4fGvTETCEAiN:b-k1kmlq) ([raw file](https://github.com/robotlocomotion/drake/blob/1b70dca12ac0ebf595f2c68a5ee1c2f9e6d75739/systems/framework/diagram_output_port.h#L36)):*
> ```objc
>   output port of one of the diagram's child subsystems.
> 
>   @param diagram The Diagram that will own this port.
> ```

BTW Capitalizing "Diagram" will cause a doxygen link to the class, but the parameter is of type `System`. This is partially resolved with your `@pre`, but not quite.  But the real reason (as provided in the comments below) is not captured in the code -- the parameter must be `System` because this class cannot downcast. I'd want to see that captured somewhere here for future developers to grok.

---

*[systems/framework/diagram_output_port.h, line 46 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDCtmgB6Mt_ryum7t8O:-LDCtmgB6Mt_ryum7t8P:b-kjgokw) ([raw file](https://github.com/robotlocomotion/drake/blob/1b70dca12ac0ebf595f2c68a5ee1c2f9e6d75739/systems/framework/diagram_output_port.h#L46)):*
> ```objc
>   @pre The `diagram` System must actually be a Diagram.
>   @pre `diagram` and `system_base` must be the same object.
>   @pre `source_output_port` must be owned by a child of this diagram.
> ```

FYI Possibly, "...by a child of `diagram`."

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8857#-:-LDCuW9P2OLDShk9CmWj:b-vm1tl1)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Both comments addressed. Also deleted a duplicate declaration I stumbled on in output_port.h (that was moved to framework_common.h long ago). PTAL.

---

Review status: 5 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[systems/framework/diagram_output_port.h, line 36 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDCsuTV4fGvTETCEAiM:-LDE0vAfAtgRqqjCIdc6:bzeczq1) ([raw file](https://github.com/robotlocomotion/drake/blob/1b70dca12ac0ebf595f2c68a5ee1c2f9e6d75739/systems/framework/diagram_output_port.h#L36)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Capitalizing "Diagram" will cause a doxygen link to the class, but the parameter is of type `System`. This is partially resolved with your `@pre`, but not quite.  But the real reason (as provided in the comments below) is not captured in the code -- the parameter must be `System` because this class cannot downcast. I'd want to see that captured somewhere here for future developers to grok.
</blockquote></details>

Done. (Just below this comment block.)

---

*[systems/framework/diagram_output_port.h, line 46 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8857#-LDCtmgB6Mt_ryum7t8O:-LDE0rHr-vUwADKBTDV4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1b70dca12ac0ebf595f2c68a5ee1c2f9e6d75739/systems/framework/diagram_output_port.h#L46)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Possibly, "...by a child of `diagram`."
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8857#-:-LDE13NvDSWuz0caREnQ:b-whlhu4)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8857)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8857)*
<!-- Sent from Reviewable.io -->
